src/Compilers/Server/ServerShared/ServerDispatcher.cs

```diff
@@ -166,32 +190,116 @@ private void WaitForAnyCompletion(IEnumerable<Task<ConnectionData>> e, Task[] ot
             }
         }
 
+        private void CreateListenTask()
+        {
+            Debug.Assert(_listenTask == null);
+            Debug.Assert(_timeoutTask == null);
+            _listenCancellationTokenSource = new CancellationTokenSource();
+            _listenTask = _clientConnectionHost.CreateListenTask(_listenCancellationTokenSource.Token);
+            _diagnosticListener.ConnectionListening();
+        }
+
+        private void CloseListenTask()
+        {
+            Debug.Assert(_listenTask != null);
+
+            _listenCancellationTokenSource.Cancel();
+            _listenCancellationTokenSource = null;
+            _listenTask = null;
+        }
+
+        private void HandleCompletedListenTask(CancellationToken cancellationToken)
+        {
+            _diagnosticListener.ConnectionReceived();
+            var allowCompilationRequests = _state == State.Running;
+            var connectionTask = HandleClientConnection(_listenTask, allowCompilationRequests, cancellationToken);
+            _connectionList.Add(connectionTask);
+
+            // Timeout and GC are only done when there are no active connections.  Now that we have a new 
+            // connection cancel out these tasks.
+            _timeoutTask = null;
+            _gcTask = null;
+
+            // Begin listening again for new connections.
+            _listenTask = null;
+            CreateListenTask();
+        }
+
+        private void HandleCompletedTimeoutTask()
+        {
+            _diagnosticListener.KeepAliveReached();
+            _listenCancellationTokenSource.Cancel();
+            _timeoutTask = null;
+            _state = State.ShuttingDown;
+        }
+
+        private void HandleCompletedGCTask()
+        {
+            _gcTask = null;
+            GC.Collect();
+        }
+
+        private void MaybeCreateTimeoutTask()
```

Comment:
"Maybe" in the name feels strange. How about changing name to CreateTimeoutTaskIfNeeded? Similar suggestion for other similar names.